### PR TITLE
OPL-22: added services for Bill/BillItem, signal to save Bill created from calcrule

### PR DESCRIPTION
**Pull request by sniedzielski**
part of ticket: https://openimis.atlassian.net/browse/OPL-22

related to the repo (branch): https://github.com/openimis/openimis-be-calcrule-third_party_payment_py/tree/feature/OPL-22

Changes: 
- added base services for Bill/BillItem and validators
- pre/post method to process creating bill in calculation rules with using added services in that PR

### CODE REVIEW

**Comment by sniedzielski** [comment]
I will add PR also for this https://github.com/openimis/openimis-be-calcrule-third_party_payment_py/tree/feature/OPL-22 when I write tests for calcrule converter (from Claims Queryset to Bill) etc.

**Review by dragos-dobre — APPROVED** [review]

File: invoice/services/bill.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    def __init__(self, user, validation_class: BillModelValidation = BillModelValidation):
+        super().__init__(user, validation_class)
+        self.validation_class = validation_class
+
+    def _base_payload_adjust(self, bill_data):
+        return self._evaluate_generic_types(bill_data)
+
+    def bill_validate_items(self, bill: Bill):
+        # TODO: Implement after calculation rules available
```
Comment (dborowiecki) [inline invoice/services/bill.py:22]: Is it still to do?

File: invoice/services/bill.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        pass
+
+    def bill_match_items(self, bill: Bill) \
+            -> Dict[str, Union[BillItemStatus, Dict[BillItem, List[BillItemStatus]]]]:
```
Comment (dborowiecki) [inline invoice/services/bill.py:26]: It's just a note, but `Union` is for X | Y.  In docstring you're more specific; we're always returning ''subject' with BillItemStatus and 'line_items' with bill items and statuses. It'd might be better idea to use `TypedDict` for this hint so it'll match docstring information.

File: invoice/signals.py [hunk 3]
```diff
@@ -52,3 +68,42 @@ def save_invoice_in_db(**kwargs):
...
             invoice_service.update(invoice_update)
+
+
+def check_bill_exist(**kwargs):
+    function_arguments = kwargs.get('data')[1]
+    instance = function_arguments.get('instance', None)
+    if instance.__class__.__name__ == "QuerySet":
+        queryset_model = instance.model
+        if queryset_model.__name__ == "Claim":
+            claim = instance.first()
```
Comment (dborowiecki) [inline invoice/signals.py:79]: Is it possible that query will return >1 entry? Maybe it'll be better to use .get() and raise exception if more than one entry is returned.

**Review by dborowiecki — APPROVED** [review]
Looks good, I left some comments.

File: invoice/services/bill.py (hunk 2 above)
Comment (sniedzielski) [reply invoice/services/bill.py:26]: Ok. I will check this. Thx for this note.

File: invoice/signals.py (hunk 3 above)
Comment (sniedzielski) [reply invoice/signals.py:79]: Yes, Could be more than 1. But this is about to check if claim is in BillItem. I can explain this details what I intended to do tomorrow

File: invoice/services/bill.py (hunk 1 above)
Comment (sniedzielski) [reply invoice/services/bill.py:22]: I think when this 1st rule regarding creating Bill is ready, then we can implement this validation.

**Review by dragos-dobre — APPROVED** [review]

**Pull request merged**